[FLINK-26028][Connector/pulsar] add sink documentation; change some source documentation (#19056)
apache/flink: docs/content/docs/connectors/datastream/pulsar.md

This part describes the Pulsar source based on the new
[data source]({{< ref "docs/dev/datastream/sources.md" >}}) API.

If you want to use the legacy `SourceFunction` or on Flink 1.13 or lower releases,
just use the StreamNative's [Pulsar Flink connector](https://github.com/streamnative/pulsar-flink).

> {{< /hint >}}

Author: Removed ~

| `persistent://sample/flink/simple-string-partition-0` | N           |
| `persistent://sample/flink/simple-string-partition-1` | N           |
| `persistent://sample/flink/simple-string-partition-2` | N           |

You can directly consume messages from the topic partitions by using the non-partitioned topic names above.
For example, use `PulsarSource.builder().setTopics("sample/flink/simple-string-partition-1", "sample/flink/simple-string-partition-2")`
would consume the partitions 1 and 2 of the `sample/flink/simple-string` topic.

#### Setting Topic Pattern

Pulsar source extracts the topic type (`persistent` or `non-persistent`) from the given topic pattern.
For example, you can use the `PulsarSource.builder().setTopicPattern("non-persistent://my-topic*")` to specify a `non-persistent` topic.
By default, a `persistent` topic is created if you do not specify the topic type in the regular expression.

You can use `setTopicPattern("topic-*", RegexSubscriptionMode.AllTopics)` to consume
both `persistent` and `non-persistent` topics based on the topic pattern.
Pulsar source would filter the available topics by the `RegexSubscriptionMode`.

### Deserializer

> A deserializer (`PulsarDeserializationSchema`) is for parsing Pulsar messages from bytes.
> You can configure the deserializer using `setDeserializationSchema(PulsarDeserializationSchema)`.

Reviewer: ```suggestion
A deserializer (`PulsarDeserializationSchema`) is for decoding Pulsar messages from bytes.
You can configure the deserializer using `setDeserializationSchema(PulsarDeserializationSchema)`.
```

#### PulsarAdmin Options

The [admin API](https://pulsar.apache.org/docs/en/admin-api-overview/) is used for querying topic metadata
and for discovering the desired topics when the Pulsar connector uses topic-pattern subscription.
It shares most part of the configuration options with the client API.
The configuration options listed here are only used in the admin API.
They are also defined in `PulsarOptions`.

{{< generated/pulsar_admin_configuration >}}

#### Pulsar Consumer Options

In general, Pulsar provides the Reader API and Consumer API for consuming messages in different scenarios.
The Pulsar connector uses the Consumer API. It extracts most parts of Pulsar's `ConsumerConfigurationData` as Flink configuration options in `PulsarSourceOptions`.

{{< generated/pulsar_consumer_configuration >}}

#### PulsarSource Options

The configuration options below are mainly used for customizing the performance and message acknowledgement behavior.
You can just leave them alone if you do not have any performance issues.

{{< generated/pulsar_source_configuration >}}

### Dynamic Partition Discovery

To handle scenarios like topic scaling-out or topic creation without restarting the Flink

> job, Pulsar source periodically discover new partitions under a provided
> topic-partition subscription pattern. To enable partition discovery, you can set a non-negative value for
> the `PulsarSourceOptions.PULSAR_PARTITION_DISCOVERY_INTERVAL_MS` option:

Reviewer: ```suggestion
job, the Pulsar source periodically discover new partitions under a provided
topic-partition subscription pattern. To enable partition discovery, you can set a non-negative value for
the `PulsarSourceOptions.PULSAR_PARTITION_DISCOVERY_INTERVAL_MS` option:
```

implementation detail and handles routing under the hood automatically. Pulsar Sink uses a lower client
API to implement its own routing layer to support multiple topics routing.

For details, see  [partitioned topics](https://pulsar.apache.org/docs/en/cookbooks-partitioned/).
{{< /hint >}}

### Delivery Guarantee

`PulsarSink` supports three delivery guarantee semantics.

- `NONE`: Data loss can happen even when the pipeline is running.
  Basically, we use a fire-and-forget strategy to send records to Pulsar topics in this mode.
  It means that this mode has the highest throughput.
- `AT_LEAST_ONCE`: No data loss happens, but data duplication can happen after a restart from checkpoint.
- `EXACTLY_ONCE`: No data loss happens. Each record is sent to the Pulsar broker only once.
  Pulsar Sink uses [Pulsar transaction](https://pulsar.apache.org/docs/en/transactions/)
  and two-phase commit (2PC) to ensure records are sent only once even after pipeline restarts.

### Delayed message delivery

[Delayed message delivery](https://pulsar.apache.org/docs/en/next/concepts-messaging/#delayed-message-delivery)

> enables you to consume a message later. With delayed message enabled,  the Pulsar sink sends a message to the Pulsar topic

Reviewer: ```suggestion
enables you to delay the possibility to consume a message. With delayed message enabled, the Pulsar sink sends a message to the Pulsar topic
```

**immediately**, but the message is delivered to a consumer once the specified delay is over.

Delayed message delivery only works in the `Shared` subscription type. In `Exclusive` and `Failover`
subscription types, the delayed message is dispatched immediately.

> You can configure the `MessageDelayer` for judging when sending the message to the consumer.

Reviewer: ```suggestion
You can configure the `MessageDelayer` to define when to send the message to the consumer.
```

> The default delayer is never delay the message dispatching. You can use the `MessageDelayer.fixed(Duration)` option to

Reviewer: ```suggestion
The default option is to never delay the message dispatching. You can use the `MessageDelayer.fixed(Duration)` option to
```

        <td>Gauge</td>
    </tr>
    <tr>
        <td>PulsarSink.producer."ProducerName".sendLatency95Pct</td>
        <td>ProducerName</td>
        <td>The 95th percentile of send latency in the last refresh interval for a specific producer.</td>
        <td>Gauge</td>
    </tr>
    <tr>
        <td>PulsarSink.producer."ProducerName".sendLatency99Pct</td>
        <td>ProducerName</td>
        <td>The 99th percentile of send latency in the last refresh interval for a specific producer.</td>
        <td>Gauge</td>
    </tr>
    <tr>
        <td>PulsarSink.producer."ProducerName".sendLatency999Pct</td>
        <td>ProducerName</td>
        <td>The 99.9th percentile of send latency in the last refresh interval for a specific producer.</td>
        <td>Gauge</td>
    </tr>
  </tbody>
</table>

{{< hint info >}}
- `numBytesOut`, `numRecordsOut`, `numRecordsOutErrors` are retrieved from Pulsar client metrics.

- `currentSendTime` tracks the time from when the producer calls `sendAync()` to
  the time when the message is acknowledged by the broker. This metric is not available in `NONE` delivery guarantee.
{{< /hint >}}

> The Pulsar producer refreshes its stats every 60 seconds by default. And PulsarSink retrieves the Pulsar producer
> stats every 500ms. Thus `numRecordsOut`, `numBytesOut`, `numAcksReceived`, and `numRecordsOutErrors` 
> is updated every 60 seconds. To increase the metrics refresh frequency, you can change
> the Pulsar producer stats refresh interval to a smaller value (minimum 1 second), as shown below.

Reviewer: ```suggestion
The Pulsar producer refreshes its stats every 60 seconds by default. The PulsarSink retrieves the Pulsar producer
stats every 500ms. That means that `numRecordsOut`, `numBytesOut`, `numAcksReceived`, and `numRecordsOutErrors` 
are updated every 60 seconds. To increase the metrics refresh frequency, you can change
the Pulsar producer stats refresh interval to a smaller value (minimum 1 second), as shown below.
```

```java
builder.setConfig(PulsarOptions.PULSAR_STATS_INTERVAL_SECONDS. 1L)
```

`numBytesOutRate` and `numRecordsOutRate` are calculated based on the `numBytesOut` and `numRecordsOUt`
counter respectively. Flink internally uses a fixed 60 seconds window to calculate the rates.

### Brief Design Rationale

Pulsar sink follow the Sink API defined in 

> [FLIP-191](https://cwiki.apache.org/confluence/display/FLINK/FLIP-191%3A+Extend+unified+Sink+interface+to+support+small+file+compaction).

Reviewer: I think you mean FLIP-143? https://cwiki.apache.org/confluence/display/FLINK/FLIP-143%3A+Unified+Sink+API

Author: Well the PulsarSInk is based on SinkV2 which is the FLIP-191 mentioned API

#### Stateless SinkWriter

> In `EXACTLY_ONCE` mode, the Pulsar sink does not store transaction information in checkpoint,
> which means a new bunch of transactions will be created after a restart.
> Therefore, any message in previous pending transactions is  either aborted or timeout
> (In no way they are visible to the downstream Pulsar consumer).
> Pulsar team is working to optimize the resource cost bought by unfinished pending transactions.

Reviewer: ```suggestion
In `EXACTLY_ONCE` mode, the Pulsar sink does not store transaction information in a checkpoint.
That means that new transactions will be created after a restart.
Therefore, any message in previous pending transactions is either aborted or timed out
(They are never visible to the downstream Pulsar consumer).
The Pulsar team is working to optimize the needed resources by unfinished pending transactions.
```

#### Pulsar Schema Evolution

[Pulsar Schema Evolution](https://pulsar.apache.org/docs/en/schema-evolution-compatibility/) allows
you to reuse the same Flink job after certain "allowed" data model changes, like adding or deleting
a field in a AVRO-based Pojo class. Please note that you can specify Pulsar schema validation rules
and define an auto schema update. For details, refer to [Pulsar Schema Evolution](https://pulsar.apache.org/docs/en/schema-evolution-compatibility/).

## Known Issues

> This section describes some known issues about the Pulsar connectors.

Reviewer: Don't we also still have a known issue with Java 11 support? I would like to have that listed too

Author: Added~

### No TransactionCoordinatorNotFound, but automatic reconnect

> Pulsar  transaction is still in active development and is not stable. Pulsar 2.9.2
> introduce [a break change](https://github.com/apache/pulsar/pull/13135) in transactions.
> If you use Pulsar 2.9.2 or higher version  with an older Pulsar client, you might get a `TransactionCoordinatorNotFound` exception.

Reviewer: ```suggestion
Pulsar transactions are still in active development and are not stable. Pulsar 2.9.2
introduces [a break change](https://github.com/apache/pulsar/pull/13135) in transactions.
If you use Pulsar 2.9.2 or higher with an older Pulsar client, you might get a `TransactionCoordinatorNotFound` exception.
```

You can use the latest `pulsar-client-all` release to resolve this issue.

## Upgrading to the Latest Connector Version

The generic upgrade steps are outlined in [upgrading jobs and Flink versions guide]({{< ref "docs/ops/upgrading" >}}).
The Pulsar connector does not store any state on the Flink side. The Pulsar connector pushes and stores all the states on the Pulsar side.
For Pulsar, you additionally need to know these limitations:

* Do not upgrade the Pulsar connector and Pulsar broker version at the same time.

> * Always use a newer Pulsar client with Pulsar connector to consuming messages from Pulsar.

Reviewer: ```suggestion
* Always use a newer Pulsar client with Pulsar connector to consume messages from Pulsar.
```

Ensure that that timeout is greater than checkpoint interval + maximum recovery time.
A shorter checkpoint interval indicates a better  consuming performance.
You can use the `PulsarSourceOptions.PULSAR_TRANSACTION_TIMEOUT_MILLIS` option to change the transaction timeout.

If checkpointing is disabled or you can not enable the transaction on Pulsar broker, you should set
`PulsarSourceOptions.PULSAR_ENABLE_AUTO_ACKNOWLEDGE_MESSAGE` to `true`.
The message is immediately acknowledged after consuming.
However, we can not promise the consistency in this scenario.

{{< hint info >}}
All acknowledgements in a transaction are recorded in the Pulsar broker side.
{{< /hint >}}

## Pulsar Sink

Pulsar Sink supports writing records into one or more Pulsar topics or a specified list of Pulsar partitions.

{{< hint info >}}
This part describes the Pulsar sink based on the new
[data sink](https://cwiki.apache.org/confluence/display/FLINK/FLIP-191%3A+Extend+unified+Sink+interface+to+support+small+file+compaction) API.

If you still want to use the legacy `SinkFunction` or on Flink 1.14 or previous releases, just use the StreamNative's
[pulsar-flink](https://github.com/streamnative/pulsar-flink).

> {{< /hint >}}

Reviewer: Like for the Source, just remove parts that are referring to older versions

Author: Sure, I'll remove it in the final commit ~

Pulsar Sink uses a builder class to construct the `PulsarSink` instance.
This example writes a String record to a Pulsar topic with at-least-once delivery guarantee.

```java
DataStream<String> stream = ...

PulsarSink<String> sink = PulsarSink.builder()
    .setServiceUrl(serviceUrl)
    .setAdminUrl(adminUrl)
    .setTopics("topic1")
    .setSerializationSchema(PulsarSerializationSchema.flinkSchema(new SimpleStringSchema()))
    .setDeliverGuarantee(DeliveryGuarantee.AT_LEAST_ONCE)
    .build();
        
stream.sinkTo(sink);
```

The following properties are **required** for building PulsarSink:

- Pulsar service url, configured by `setServiceUrl(String)`
- Pulsar service http url (aka. admin url), configured by `setAdminUrl(String)`
- Topics / partitions to write, see [writing targets](#writing-targets) for more details.
- Serializer to generate Pulsar messages, see [serializer](#serializer) for more details.

It is **recommended** to set the producer name in Pulsar Sink by `setProducerName(String)`.
This gives a unique name to the flink connector in the Pulsar statistic dashboard.
Give you a mechanism to monitor the performance of your flink applications.

### Writing Topics

> Setting the topics to write is a bit like the [topic-partition subscription](#topic-partition-subscription)
> in Pulsar source. We support a mixin style of topic setting. Therefore, you can provide a list of topics,
> partitions, or both of them.

Reviewer: ```suggestion
Defining the topics for producing is similar to the [topic-partition subscription](#topic-partition-subscription)
in the Pulsar source. We support a mix-in style of topic setting. You can provide a list of topics,
partitions, or both of them.
```

```java
// Topic "some-topic1" and "some-topic2"
PulsarSink.builder().setTopics("some-topic1", "some-topic2")

// Partition 0 and 2 of topic "topic-a"
PulsarSink.builder().setTopics("topic-a-partition-0", "topic-a-partition-2")

// Partition 0 and 2 of topic "topic-a" and topic "some-topic2"
PulsarSink.builder().setTopics("topic-a-partition-0", "topic-a-partition-2", "some-topic2")
```

The topics you provide support auto partition discovery. We query the topic metadata from the Pulsar in a fixed interval.
You can use the `PulsarSinkOptions.PULSAR_TOPIC_METADATA_REFRESH_INTERVAL` option to change the discovery interval option.

Configure writing targets can be replaced by using a custom [`TopicRouter`]
[message routing](#message-routing). And read [flexible topic naming](#flexible-topic-naming)
for understanding how to configure partitions on the Pulsar connector.

{{< hint warning >}}
If you build  the Pulsar sink based on both the topic and its corresponding partitions, Pulsar sink merges them and only use the topic.

> For example, it uses the `PulsarSink.builder().setTopics("some-topic1", "some-topic1-partition-0")` option to build the Pulsar sink,
> it is simplified to `PulsarSink.builder().setTopics("some-topic1")`.

Reviewer: ```suggestion
For example, when using the `PulsarSink.builder().setTopics("some-topic1", "some-topic1-partition-0")` option to build the Pulsar sink,
this is simplified to `PulsarSink.builder().setTopics("some-topic1")`.
```

{{< /hint >}}

### Serializer

A serializer (`PulsarSerializationSchema`) is required to serializing the record instance into bytes.
Similar to `PulsarSource`, Pulsar sink supports both Flink's `SerializationSchema` and
Pulsar's `Schema`. But Pulsar's `Schema.AUTO_PRODUCE_BYTES()` is not supported in Pulsar Sink.

If you do not need the message key and other message properties in Pulsar's

> [Message](https://pulsar.apache.org/api/client/2.9.0-SNAPSHOT/org/apache/pulsar/client/api/Message.html) interface,

Author: This is my mistake though this refers to a concept that is pretty stable. I'l; change it to use the latest stable version.

> [Message](https://pulsar.apache.org/api/client/2.9.0-SNAPSHOT/org/apache/pulsar/client/api/Message.html) interface,

Reviewer 2: Pulsar use `SNAPSHOT` javadocs. There is no stable version. I'm afraid we have to use the snapshot here.

> you can use the predefined `PulsarSerializationSchema`. Pulsar sink provides two implementation methods.

Reviewer: ```suggestion
you can use the predefined `PulsarSerializationSchema`. The Pulsar sink provides two implementation methods.
```
